Restore states for RFLink devices

javicalle opened the pull request:
## Description:
This PR aims to add the ability to recover the previous state to RFLink devices.
RFLink devices do not have the ability to ask for their status, so in the reboots, the previous state is lost. This delivery gives these devices the ability to 'remember' the status on the reboots.
This PR comes from the failed PR #18208

Some of the points to review:
- I don't think that will be a breaking change
- If the PR goes live I will adjust the documentation to the new behavior (on reboots)
- Sugestions and improvement proposals

My intention would be to release an initial version 'as is' and then work on some of the points of improvement:
1. How to disable the behavior? My approach would be to make use of a new 'initial_value' attribute that, if informed, prevails over the previous value. I don't like the option to define a flag attribute to modify the behavior and I think that 'initial_value' is already used in other cases.
2. The new behavior does not apply to devices of the `sensor` and `binary_sensor` type. In my opinion, I do not believe that the previous state should be restored in these cases, in the next update the correct state will be restored, regardless of the previous state.

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#<home-assistant.io PR number goes here>

## Example entry for `configuration.yaml` (if applicable):
```yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]). **N/A**
  - New dependencies are only imported inside functions that use them ([example][ex-import]). **N/A**
  - New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`. **N/A**
  - New files were added to `.coveragerc`. **N/A**

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

javicalle committed:
Merge branch 'master' of https://github.com/home-assistant/home-assistant into dev

# Conflicts:
#	homeassistant/components/binary_sensor/point.py
#	homeassistant/components/cloud/__init__.py
#	homeassistant/components/cloud/prefs.py
#	homeassistant/components/frontend/__init__.py
#	homeassistant/components/light/fibaro.py
#	homeassistant/components/logbook.py
#	homeassistant/components/point/__init__.py
#	homeassistant/config_entries.py
#	homeassistant/const.py
#	homeassistant/helpers/service.py
#	requirements_all.txt
#	requirements_test_all.txt

javicalle committed:
one 'async_get_last_state' refactor left behind

javicalle committed:
Remove RestoreEntity inheritance (already in parent class)

javicalle commented:
Well, after the `RestoreEntity` refactor, now I am having a few `Too many ancestors` pylint errors:
```
************* Module homeassistant.components.switch.rflink
homeassistant/components/switch/rflink.py:70:0: R0901: Too many ancestors (8/7) (too-many-ancestors)
************* Module homeassistant.components.light.rflink
homeassistant/components/light/rflink.py:152:0: R0901: Too many ancestors (8/7) (too-many-ancestors)
homeassistant/components/light/rflink.py:158:0: R0901: Too many ancestors (8/7) (too-many-ancestors)
homeassistant/components/light/rflink.py:192:0: R0901: Too many ancestors (8/7) (too-many-ancestors)
homeassistant/components/light/rflink.py:244:0: R0901: Too many ancestors (8/7) (too-many-ancestors)
```

For all the other Travis errors, I will take a look when I have a break.

I think that refactor the inheritance for RFLink classes wil be a delicate task that I do not know if I can/must assume.
Code 'owner' has commented on several occasions that he can not devote time to it.

I will see what can I do.

emontnemery commented:
@javicalle You can disable the warning like this:
``` python
# pylint: disable=too-many-ancestors
class MqttSwitch(MqttAvailability, MqttDiscoveryUpdate, MqttEntityDeviceInfo,
                 SwitchDevice, RestoreEntity):
```

javicalle committed:
# pylint: disable=too-many-ancestors

javicalle committed:
code predictor can be a bitch

javicalle committed:
lint corrections

javicalle committed:
# pylint: disable=too-many-ancestors

pvizeli commented:
Can you rebase your PR? There was a change in background and verify that it works with new handling?

javicalle commented:
> Can you rebase your PR? There was a change in background and verify that it works with new handling?

@pvizeli Done.
Anything else I can help?

javicalle commented on tests/components/cover/test_rflink.py:
```diff
@@ -0,0 +1,501 @@
+    assert protocol.send_command_ack.call_count == 2
+
+    # test if default apply to configured devices
+    hass.async_add_job(
```
Done

javicalle commented on tests/components/cover/test_rflink.py:
```diff
@@ -0,0 +1,501 @@
+        hass, config, DOMAIN, monkeypatch)
+
+    # test if signal repetition is performed according to configuration
+    hass.async_add_job(
```
Done

javicalle commented on tests/components/cover/test_rflink.py:
```diff
@@ -0,0 +1,501 @@
+    assert protocol.send_command_ack.call_args_list[0][0][0] == 'protocol_0_0'
+    assert protocol.send_command_ack.call_args_list[0][0][1] == 'DOWN'
+
+    hass.async_add_job(
```
Done

javicalle commented on tests/components/cover/test_rflink.py:
```diff
@@ -0,0 +1,501 @@
+    cover_after_first_command = hass.states.get(DOMAIN + '.test')
+    assert cover_after_first_command.state == STATE_OPEN
+
+    # should repond to group command
```
Upsss

javicalle commented on tests/components/cover/test_rflink.py:
```diff
@@ -0,0 +1,501 @@
+    assert hass.states.get(DOMAIN + '.test').state == STATE_OPEN
+
+    # test changing state from HA propagates to RFLink
+    hass.async_add_job(
```
Done

javicalle commented on tests/components/cover/test_rflink.py:
```diff
@@ -0,0 +1,501 @@
+def test_default_setup(hass, monkeypatch):
+    """Test all basic functionality of the RFLink cover component."""
+    # setup mocking rflink module
+    event_callback, create, protocol, _ = yield from mock_rflink(
```
Changed in all the calls

javicalle commented on tests/components/cover/test_rflink.py:
```diff
@@ -0,0 +1,501 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+@asyncio.coroutine
```
Done

javicalle commented on homeassistant/components/light/rflink.py:
```diff
@@ -148,17 +148,29 @@ def devices_from_config(domain_config):
+        if old_state is not None and \
+                old_state.attributes.get(ATTR_BRIGHTNESS) is not None:
+            # restore also brightness in dimmables devices
+            self._brightness = int(old_state.attributes.get(ATTR_BRIGHTNESS))
```
If I have understood you correctly, done

javicalle commented on homeassistant/components/light/rflink.py:
```diff
@@ -196,6 +209,16 @@ class HybridRflinkLight(SwitchableRflinkDevice, Light):
+        if old_state is not None and \
+                old_state.attributes.get(ATTR_BRIGHTNESS) is not None:
+            # restore also brightness in dimmables devices
+            self._brightness = int(old_state.attributes.get(ATTR_BRIGHTNESS))
```
If I have understood you correctly, done

javicalle committed:
recover from dict[key]

javicalle committed:
Remove all 'coroutine' decorator, replace for 'async def'
Replace all 'yield from' for 'await'
Replace 'hass.async_add_job' for 'hass.async_create_task'

MartinHjelmare reviewed (approved):
Looks good! @emontnemery what do you think?

emontnemery commented:
I also think it looks good!

javicalle commented:
Thanks!
